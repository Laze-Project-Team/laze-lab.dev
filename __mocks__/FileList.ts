class FileList<T> {
  _files: T[];
  length: number;
  constructor(files: T[]) {
    this._files = files;
    this.length = this._files.length;
  }
  item(index: number) {
    return this._files[index];
  }
  get [Symbol.toStringTag]() {
    return 'FileList';
  }
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
window.FileList = FileList;

export {};
