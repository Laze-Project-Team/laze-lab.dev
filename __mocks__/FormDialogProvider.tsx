import type { FC, ReactNode } from 'react';
import { useForm } from 'react-hook-form';

import { FormDialogContextProvider } from '@/components/ui/FormDialog/FormDialogContext';

export const FormDialogProvider: FC<{ children?: ReactNode }> = ({
  children,
}) => {
  const { control } = useForm();
  return (
    <FormDialogContextProvider
      value={{
        id: 'mock-form-dialog',
        control,
        onSubmit: async (data) => console.log(data),
      }}
    >
      {children}
    </FormDialogContextProvider>
  );
};
