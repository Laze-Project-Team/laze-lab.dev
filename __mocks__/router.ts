import type { useRouter } from 'next/router';
import { beforeEach, vi } from 'vitest';

export const mockRouter = (): void => {
  vi.mock('next/router', () => ({
    useRouter: () => {
      const router: ReturnType<typeof useRouter> = {
        asPath: '/',
        back: vi.fn(),
        basePath: '/',
        beforePopState: vi.fn(),
        events: {
          on: vi.fn(),
          off: vi.fn(),
          emit: vi.fn(),
        },
        isFallback: false,
        isLocaleDomain: false,
        isPreview: false,
        isReady: true,
        locale: 'en',
        locales: ['en'],
        prefetch: vi.fn(),
        push: vi.fn(),
        query: {},
        reload: vi.fn(),
        replace: vi.fn(),
        route: '/',
        pathname: '/',
      };

      return router;
    },
  }));
};

export const mockEachRouter = (): void => {
  beforeEach(() => {
    mockRouter();
  });
};
