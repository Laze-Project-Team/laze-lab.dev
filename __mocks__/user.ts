import type { User } from 'firebase/auth';

import type { userData } from '@/typings/database';

export const userMock: User | null | undefined = {
  uid: 'unique-id',
  delete: async () => void 0,
  displayName: 'Laze Project Team',
  email: 'laze@example.com',
  emailVerified: true,
  getIdToken: async () => 'token',
  getIdTokenResult: async () => ({
    authTime: new Date().toISOString(),
    expirationTime: new Date().toISOString(),
    issuedAtTime: new Date().toISOString(),
    signInProvider: null,
    signInSecondFactor: null,
    token: 'token',
    claims: {},
  }),
  isAnonymous: false,
  metadata: {},
  phoneNumber: null,
  photoURL: null,
  providerData: [],
  providerId: 'email',
  refreshToken: 'token',
  reload: async () => void 0,
  tenantId: null,
  toJSON: () => ({}),
};

export const userDataMock: userData | null | undefined = {
  name: 'Laze Project Team',
  avatarURL: 'https://laze.ddns.net/img/logo/logo.png',
  locale: 'ja',
  projects: [],
  createdAt: new Date(),
};
