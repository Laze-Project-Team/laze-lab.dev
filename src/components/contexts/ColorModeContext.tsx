import type { PaletteMode, Theme } from '@mui/material';
import { CssBaseline, ThemeProvider, useMediaQuery } from '@mui/material';
import type { FC, ReactNode } from 'react';
import { createContext, useContext, useEffect, useMemo, useState } from 'react';

import { getTheme } from '@/lib/theme';

export type colorModeContextType = {
  toggleColorMode: () => void;
  colorMode: PaletteMode;
  theme: Theme;
};

const colorModeContext = createContext<colorModeContextType | null>(null);

export const useColorMode = (): colorModeContextType => {
  const colorMode = useContext(colorModeContext);

  if (colorMode === null) {
    throw new Error('Please wrap component with ColorModeProvider');
  }

  return colorMode;
};

export type colorModeProviderProps = {
  children?: ReactNode;
  preferTheme?: PaletteMode;
  isLocal?: boolean;
};

export const ColorModeProvider: FC<colorModeProviderProps> = ({
  children,
  preferTheme,
  isLocal,
}) => {
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');
  const [mode, setMode] = useState<PaletteMode>(
    preferTheme ?? prefersDarkMode ? 'dark' : 'light',
  );

  useEffect(() => {
    if (preferTheme !== undefined) {
      setMode(preferTheme);
    }
  }, [preferTheme]);

  // Update the theme only if the mode changes
  const theme = useMemo(() => getTheme(mode), [mode]);

  const colorMode = useMemo(
    () => ({
      // The dark mode switch would invoke this method
      toggleColorMode: () => {
        setMode((prevMode) => (prevMode === 'light' ? 'dark' : 'light'));
      },
      colorMode: mode,
      theme,
    }),
    [mode, theme],
  );

  return (
    <colorModeContext.Provider value={colorMode}>
      <ThemeProvider theme={theme}>
        {!isLocal && <CssBaseline />}
        {children}
      </ThemeProvider>
    </colorModeContext.Provider>
  );
};
