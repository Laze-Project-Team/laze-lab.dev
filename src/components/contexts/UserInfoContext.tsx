import type { FC, ReactNode } from 'react';
import { createContext, useContext } from 'react';

import type { userInfo } from '@/components/hooks/useUserInfo';
import { useUserInfoStrictly } from '@/components/hooks/useUserInfo';

export const userInfoContext = createContext<userInfo | null>(null);

export const useUserInfoContext = (): userInfo => {
  const userInfo = useContext(userInfoContext);

  if (userInfo === null) {
    throw new Error('please wrap component with UserInfoProvider');
  }

  return userInfo;
};

export const UserInfoProvider: FC<{
  children: ReactNode;
}> = ({ children }) => {
  const userInfo = useUserInfoStrictly();

  return (
    <userInfoContext.Provider value={userInfo}>
      {children}
    </userInfoContext.Provider>
  );
};
