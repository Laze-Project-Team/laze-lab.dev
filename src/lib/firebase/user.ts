import type { User } from 'firebase/auth';
import { onAuthStateChanged } from 'firebase/auth';
import type { UpdateData } from 'firebase/firestore';
import { deleteDoc, doc, getDoc, setDoc, updateDoc } from 'firebase/firestore';
import { getDownloadURL, ref, uploadBytes } from 'firebase/storage';
import { useEffect } from 'react';
import type { SWRResponse } from 'swr';
import { mutate } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { timestampToDate } from '@/lib/firebase/TimestampToDate';
import { isLocale } from '@/lib/utils/isLocale';
import type { baseUserData, userData } from '@/typings/database';

import { useFirebase } from '.';

export type userManager = {
  getUser: () => SWRResponse<User | null | undefined>;
  syncUser: () => void;
  createUser: (user: User) => Promise<userData>;
  fetchUserData: (user: User) => Promise<userData>;
  updateUserData: (user: User, data: UpdateData<baseUserData>) => Promise<void>;
  updateUserAvatar: (user: User, avatarFile: File) => Promise<void>;
  deleteUser: (user: User) => Promise<void>;
};

export const useUserManager = (): userManager => {
  const { auth, db, storage } = useFirebase();
  const currentUser = useSWRImmutable<User | null | undefined>(
    'currentUser',
    () => auth.currentUser,
  );

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      mutate('currentUser', user);
    });

    return unsubscribe;
  }, [auth]);

  const getUser: userManager['getUser'] = () => {
    return currentUser;
  };

  const syncUser: userManager['syncUser'] = () => {
    mutate('currentUser', undefined);
    mutate('currentUser', auth.currentUser);
  };

  const createUser: userManager['createUser'] = async (user) => {
    const userData: userData = {
      avatarURL: user.photoURL,
      locale: isLocale(auth.languageCode) ? auth.languageCode : 'en',
      name: user.displayName ?? 'user',
      projects: [],
      createdAt: new Date(),
    };

    await setDoc(doc(db, 'users', user.uid), userData);

    return userData;
  };

  const fetchUserData: userManager['fetchUserData'] = async (user) => {
    const snapshot = await getDoc(doc(db, 'users', user.uid));

    if (!snapshot.exists()) {
      return await createUser(user);
    }

    const userData = timestampToDate(snapshot.data()) as userData;

    return userData;
  };

  const updateUserData: userManager['updateUserData'] = async (user, data) => {
    await updateDoc(doc(db, 'users', user.uid), data);
  };

  const updateUserAvatar: userManager['updateUserAvatar'] = async (
    user,
    avatarFile,
  ) => {
    const avatarFileRef = ref(storage, `users/${user.uid}`);

    const { ref: uploadedRef } = await uploadBytes(avatarFileRef, avatarFile);

    const avatarURL = await getDownloadURL(uploadedRef);

    await updateUserData(user, { avatarURL });
  };

  const deleteUser: userManager['deleteUser'] = async (user) => {
    await deleteDoc(doc(db, 'users', user.uid));
  };

  return {
    getUser,
    syncUser,
    createUser,
    fetchUserData,
    updateUserData,
    updateUserAvatar,
    deleteUser,
  };
};
