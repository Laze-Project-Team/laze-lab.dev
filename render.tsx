import type {
  Queries,
  queries,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { render as defaultRender } from '@testing-library/react';
import type { FC, ReactNode } from 'react';

import { ColorModeProvider } from '@/components/contexts/ColorModeContext';

export type renderType<
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container,
> = (
  ui: React.ReactElement,
  options?: RenderOptions<Q, Container, BaseElement>,
) => RenderResult<Q, Container, BaseElement>;

type wrapperProps = {
  children?: ReactNode;
};

type globalWrapperHOCType = (
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>,
) => FC<wrapperProps>;

const globalWrapperHOC: globalWrapperHOCType = (Wrapper) => {
  const globalWrapper: FC<wrapperProps> = ({ children }) => (
    <ColorModeProvider>
      {Wrapper ? (
        <Wrapper>
          <>{children}</>
        </Wrapper>
      ) : (
        children
      )}
    </ColorModeProvider>
  );

  return globalWrapper;
};

export const render: renderType = (ui, options) =>
  defaultRender(ui, {
    ...options,
    wrapper: globalWrapperHOC(options?.wrapper),
  });
