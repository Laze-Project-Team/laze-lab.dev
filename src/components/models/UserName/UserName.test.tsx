import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import { expect, it } from 'vitest';

import { userDataMock, userMock } from '/__mocks__/user';
import { userInfoContext } from '@/components/contexts/UserInfoContext';
import { render } from '@/lib/test/render';

import type { userNameProps } from '.';
import { UserName } from '.';

it('should be rendered correctly', () => {
  const { container } = render(<UserName />);
  expect(container.firstChild).toMatchSnapshot();
});

it('should render username', () => {
  const username = "I'm a user";
  render(
    <userInfoContext.Provider
      value={{
        user: userMock,
        userData: {
          ...userDataMock,
          data: { ...userDataMock.data, name: username },
        } as userNameProps['userData'],
        syncUserData: () => void 0,
      }}
    >
      <UserName />
    </userInfoContext.Provider>,
  );

  expect(screen.getByText(username)).toBeInTheDocument();
});
