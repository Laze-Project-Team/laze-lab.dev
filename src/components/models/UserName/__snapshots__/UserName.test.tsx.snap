// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be rendered correctly 1`] = `
<div>
  <p
    class="css-ai2zs0-PresentialUserName"
  >
    Laze Project Team
  </p>
</div>
`;
